src/script/sign.h

```diff
@@ -67,9 +67,238 @@ struct SignatureData {
+            WriteCompactSize(s, entry.first.size() + 1);
+            s << (char)0x01;
+            s << entry.first;
+            WriteCompactSize(s, entry.second.size());
+            s << CFlatData(REF(entry.second));
+        }
+        for (auto& entry : witness_scripts) {
+            WriteCompactSize(s, entry.first.size() + 1);
+            s << (char)0x02;
+            s << entry.first;
+            WriteCompactSize(s, entry.second.size());
+            s << CFlatData(REF(entry.second));
+        }
+
+        // Separator
+        s << (char)0xff;
+
+        for (unsigned int i = 0; i < tx.vin.size(); i++) {
+            CTxIn in = tx.vin[i];
+            if (in.scriptSig.empty() && in.scriptWitness.IsNull()) {
+                // Write the utxo
+                auto non_witness_utxo_it = non_witness_utxos.find(in.prevout);
+                auto witness_utxo_it = witness_utxos.find(in.prevout);
+                // If there is a witnes utxo, then don't add the nonwitnes one.
+                if (witness_utxo_it != witness_utxos.end()) {
+                    s << (char)0x01 << (char)0x04;
+                    WriteCompactSize(s, ::GetSerializeSize(witness_utxo_it->second, SER_NETWORK, PROTOCOL_VERSION));
+                    s << witness_utxo_it->second;
+                } else if (non_witness_utxo_it != non_witness_utxos.end()) {
+                    s << (char)0x01 << (char)0x03;
+                    WriteCompactSize(s, ::GetSerializeSize(non_witness_utxo_it->second, SER_NETWORK, PROTOCOL_VERSION));
+                    s << non_witness_utxo_it->second;
+                }
+
+                // Write any partial signatures
+                auto partial_sig = partial_sigs.at(i);
+                for (auto sig_pair : partial_sig) {
+                    WriteCompactSize(s, sig_pair.first.size() + 1);
+                    s << (char)0x05;
+                    s << CFlatData((void *)REF(sig_pair.first.begin()), (void *)REF(sig_pair.first.end()));
+                    WriteCompactSize(s, sig_pair.second.size());
+                    s << CFlatData(REF(sig_pair.second));
+                }
+            }
+            s << (char)0xff;
+        }
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read the magic bytes
+        unsigned char magic[5];
+        CFlatData magic_data(magic, magic + 5);
+        s >> magic_data;
+
+        // Expect types to only increase sequentually, so remember last type
+        unsigned char last_type = 0x00;
+
+        // hashers
+        CHash160 hash160_hasher;
+        CSHA256 sha256_hasher;
+
+        // Read loop
+        int separators = 0;
+        std::map<CPubKey, std::vector<unsigned char>> inputs_partial_sigs;
+        while(!s.empty()) {
+            // Reset hashers
+            hash160_hasher.Reset();
+            sha256_hasher.Reset();
+
+            // Check for separator
+            unsigned char next_byte = s[0];
```

Comment:
I don't expect [0] this to work for every Stream implementation. What about instead trying the separator as a special key? In particular, you could make the "" key (which is serialized as just 0x00) the separator. Then your loop can read keys one by one, until it encounters this special value, and bail out.